Feature: Add a persistent backend to the cloud-init server which uses duckdb via quack/quack

alexlovelltroy opened the pull request:
This pull request introduces a new storage backend, `quackstore`, and includes various updates to support this addition. The most significant changes involve modifying the main application logic to handle the new backend, adding the necessary dependencies, and implementing and testing the `quackstore` functionality.

### Main Application Updates:
* Added `quackstore` import and new flags for selecting the storage backend and database path in `cmd/cloud-init-server/main.go`. [[1]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918R18) [[2]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918R54-R55) [[3]](diffhunk://#diff-b0d51809eac0491964f1e5e9536026b83ba4c1036480c4143cebdfa7cd4eb918R75-R76)
* Updated the main function to initialize the datastore based on the selected backend (`memstore` or `quackstore`).

### New `quackstore` Implementation:
* Implemented `quackstore` package with methods for managing groups, instances, and cluster defaults, including initialization and schema setup.

### Testing Enhancements:
* Added a test for `memstore` to ensure it adheres to the standard test suite.
* Added tests for `quackstore` to ensure proper functionality and integration.

### Model Serialization:
* Enhanced `CloudConfigFile` struct to implement custom JSON marshaling and unmarshaling, handling different encoding types (`base64` and `plain`).

synackd commented:
Any progress on this?

synackd commented:
Also:
- Looks like there have been some new versions since the last merge from `main` that we'll probably want to incorporate.
- Looks like DCO check is failing.

alexlovelltroy commented:
@synackd This PR adds two flags to control the behavior of the storage backend.
```
      --db-path string            Path to the database file for quackstore backend (default "cloud-init.db")
      --storage-backend string    Storage backend to use (mem or quack) (default "mem")
```

In addition to running the unit tests in the `internal/quackstore` directory, you can start the cloud-init server with these flags and use it normally.  To test the durable storage, feel free to `kill -9 cloud-init-server` and then restart.  It should retain any information that was loaded previously.

In a containerized deployment, the files should be stored in a volume that is remounted correctly when restarting the container.  It is easier to test running manually though.

synackd commented:
Building via goreleaser yields the same thing.

alexlovelltroy commented:
Cross-platform builds with CGO are easier on ubuntu and should require a glibc container rather than a musl one.  I've updated the container and added a build-in-container script that should replicate the way github will do the builds.

synackd commented:
With the most recent commit, I get the following error in Goreleaser when running `goreleaser release --clean --snapshot --skip publish,archive`:

```
  ⨯ release failed after 15s                        
    error=
    │ docker build failed: failed to build ghcr.io/openchami/cloud-init:v1.2.3-amd64: exit status 1: STEP 1/10: FROM registry.access.redhat.com/ubi9/ubi-minimal
    │ Trying to pull registry.access.redhat.com/ubi9/ubi-minimal:latest...
    │ Getting image source signatures
    │ Checking if image destination supports signatures
    │ Copying blob sha256:e7dc8c72147da3785cb494f592a4596697015bf12fb34fe8b2bf842d6105b702
    │ Copying config sha256:04e3acee198ee12f9f16a6f99ff124d7cc0d7e73971f2386e0cf353717cac13d
    │ Writing manifest to image destination
    │ Storing signatures
    │ STEP 2/10: RUN microdnf install -y         libstdc++         libgcc         curl         iproute         iputils         wireguard-tools         tini     && microdnf clean all
    │ (microdnf:2): librhsm-WARNING **: 23:17:41.323: Found 0 entitlement certificates
    │ (microdnf:2): librhsm-WARNING **: 23:17:41.324: Found 0 entitlement certificates
    │ Downloading metadata...
    │ Downloading metadata...
    │ Downloading metadata...
    │ error: No package matches 'wireguard-tools'
    │ Error: building at STEP "RUN microdnf install -y         libstdc++         libgcc         curl         iproute         iputils         wireguard-tools         tini     && microdnf clean all": while running runtime: exit status 1
    │ Learn more at https://goreleaser.com/errors/docker-build
```

alexlovelltroy commented:
Build automation is noble yak shaving...  The yak never gets any smaller.  At least this time I think the builds are happy.

synackd reviewed (changes requested):
I finally got the container to build with the script after applying the suggested edits.

However, when starting the container, I see:

```
cloud-init-server[2766863]: /usr/local/bin/cloud-init-server: /lib64/libm.so.6: version `GLIBC_2.38' not found (required by /usr/local/bin/cloud-init-server)
cloud-init-server[2766863]: /usr/local/bin/cloud-init-server: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /usr/local/bin/cloud-init-server)
```

Run running the binary itself via

```
./cloud-init-server --storage-backend quack --db-path ./cloud-init.db
```

it panics:

```
No JWKS URL provided; secure route will be disabled
5:32PM ERR Failed to get SMD data error="Get \"http://smd:27779/hsm/v2/Inventory/EthernetInterfaces/\": dial tcp: lookup smd: no such host"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xab17f4]

goroutine 1 [running]:
github.com/go-chi/chi/v5.chain({0xc00038a0c8, 0x1, 0x2779680?}, {0x2925720?, 0x281f2a0?})
        /root/go/pkg/mod/github.com/go-chi/chi/v5@v5.2.1/chain.go:43 +0x34
github.com/go-chi/chi/v5.Middlewares.Handler(...)
        /root/go/pkg/mod/github.com/go-chi/chi/v5@v5.2.1/chain.go:13
github.com/go-chi/chi/v5.(*Mux).handle(0xc0003842a0, 0x8, {0x27ca46e, 0xa}, {0x2925720, 0x281f2a0})
        /root/go/pkg/mod/github.com/go-chi/chi/v5@v5.2.1/mux.go:436 +0x1ae
github.com/go-chi/chi/v5.(*Mux).Get(0xc000384240?, {0x27ca46e?, 0x1?}, 0x735e79?)
        /root/go/pkg/mod/github.com/go-chi/chi/v5@v5.2.1/mux.go:162 +0x2d
main.initCiClientRouter({0x293f9c0, 0xc000384240}, 0xc000051c40, 0x0)
        /opt/shared/cloud-init/cmd/cloud-init-server/main.go:265 +0xeb
main.startServer()
        /opt/shared/cloud-init/cmd/cloud-init-server/main.go:228 +0xbeb
main.main.func1(0xc000058400?, {0x27c4c6e?, 0x4?, 0x27c4b4a?})
        /opt/shared/cloud-init/cmd/cloud-init-server/main.go:65 +0xf
github.com/spf13/cobra.(*Command).execute(0xc0001d5b08, {0xc000034060, 0x4, 0x4})
        /root/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1015 +0xa94
github.com/spf13/cobra.(*Command).ExecuteC(0xc0001d5b08)
        /root/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1148 +0x40c
github.com/spf13/cobra.(*Command).Execute(...)
        /root/go/pkg/mod/github.com/spf13/cobra@v1.9.1/command.go:1071
main.main()
        /opt/shared/cloud-init/cmd/cloud-init-server/main.go:75 +0x6f
```

Though this doesn't seem to be related to this pull request. Applying the following fix:

```diff
diff --git a/cmd/cloud-init-server/main.go b/cmd/cloud-init-server/main.go
index 9b5113a..9cb98f6 100644
--- a/cmd/cloud-init-server/main.go
+++ b/cmd/cloud-init-server/main.go
@@ -262,10 +262,17 @@ func initCiClientRouter(router chi.Router, handler *CiHandler, wgInterfaceManage
        // Add cloud-init endpoints to router
        router.Get("/openapi.json", DocsHandler)
        router.Get("/version", VersionHandler)
-       router.With(wireGuardMiddleware).Get("/user-data", UserDataHandler)
-       router.With(wireGuardMiddleware).Get("/meta-data", MetaDataHandler(handler.sm, handler.store))
-       router.With(wireGuardMiddleware).Get("/vendor-data", VendorDataHandler(handler.sm, handler.store))
-       router.With(wireGuardMiddleware).Get("/{group}.yaml", GroupUserDataHandler(handler.sm, handler.store))
+       if wireGuardMiddleware != nil {
+               router.With(wireGuardMiddleware).Get("/user-data", UserDataHandler)
+               router.With(wireGuardMiddleware).Get("/meta-data", MetaDataHandler(handler.sm, handler.store))
+               router.With(wireGuardMiddleware).Get("/vendor-data", VendorDataHandler(handler.sm, handler.store))
+               router.With(wireGuardMiddleware).Get("/{group}.yaml", GroupUserDataHandler(handler.sm, handler.store))
+       } else {
+               router.Get("/user-data", UserDataHandler)
+               router.Get("/meta-data", MetaDataHandler(handler.sm, handler.store))
+               router.Get("/vendor-data", VendorDataHandler(handler.sm, handler.store))
+               router.Get("/{group}.yaml", GroupUserDataHandler(handler.sm, handler.store))
+       }
        router.Post("/phone-home/{id}", PhoneHomeHandler(wgInterfaceManager, handler.sm))
        router.Post("/wg-init", wgtunnel.AddClientHandler(wgInterfaceManager, handler.sm))
 }
```
and I was finally able to test! :)

I did some very basic tests and I was able to persistently store cluster-defaults and group data.

synackd commented:
Glibc version errors are still occurring in the container. This is because `ubuntu:24.04` is being used to build the cloud-init binary, which is then copied into a `rockylinux:9` container to be run. Since `CGO_ENABLED=1`, the binary depends on a specific Glibc version, one that differs between the two containers.

I think the solution is to build in the same container which cloud-init will run, `rockylinux:9`. I have gotten a successful build working on my branch here: [`synackd/container-build-fixes`](https://github.com/OpenCHAMI/cloud-init/tree/synackd/container-build-fixes). It adds a separate `.goreleaser-buildincontainer.yaml` file that sets appropriate linker variables for ARM, as well as a separate `build-in-container-r9.sh` build script. These were added as separate files so that the differences can be seen with the existing ones.

With docker:

```
./build-in-container-r9.sh
```

With Podman (what I've tested with, Podman socket needs to be enable):

```
CONTAINER_CMD=podman CONTAINER_SOCK=/run/podman/podman.sock ./build-in-container-r9.sh
```

synackd commented:
Now that I've gotten the container running, I've tried setting:
```bash
STORAGE_BACKEND=quack
DB_PATH=/cloud-init/cloud-init.db
DEBUG=true
```
and am using the following quadlet with an existing OpenCHAMI installation (the `cloud-init` volume is a simple Podman volume with no custom config):
```ini
[Unit]
Description=The cloud-init-server container
Wants=smd.service
After=smd.service opaal.service
PartOf=openchami.target

[Container]
ContainerName=cloud-init-server
HostName=cloud-init
Image=ghcr.io/openchami/cloud-init:v1.2.3-amd64

Volume=cloud-init:/cloud-init:rw,Z

# Environment Variables
EnvironmentFile=/etc/openchami/configs/openchami.env

# Networks for the Container to use
Network=openchami-internal.network

# Proxy settings
PodmanArgs=--http-proxy=false

[Service]
Restart=always
```
I can see the DB path and enablement get picked up, but DuckDB is attempting to create a directory in the root directory at `/.duckdb`, which errs due to a permissions error:
```
cloud-init-server[426685]: {"level":"debug","listen":":27777","token-url":"http://opaal:3333/token","smd-url":"http://smd:27779","jwks-url":"http://opaal:3333/keys","access-token":"","cluster-name":"","region":"","az":"","cloud-provider":"","base-url":"","cacert":"","insecure":false,"impersonation":true,"wireguard-server":"","wireguard-only":false,"debug":true,"storage-backend":"quack","db-path":"/cloud-init/cloud-init.db","time":"2025-08-15T23:17:38Z","message":"Resolved configuration"}
cloud-init-server[426685]: 11:17PM ERR Failed to load DuckDB extensions error="IO Error: Failed to create directory \"/.duckdb/\": Permission denied"
```
This seems to be occurring in the `loadExtensions()` function [here](https://github.com/OpenCHAMI/quack/blob/8a577c49108460c7a1d4a709966ece78efa2706a/quack/main.go#L92). From what I can tell, the path set in `DB_PATH` gets propagated to the DB creation, so I'm not sure where this is occurring or how it's choosing where to write.

synackd commented:
> Glibc version errors are still occurring in the container. This is because `ubuntu:24.04` is being used to build the cloud-init binary, which is then copied into a `rockylinux:9` container to be run. Since `CGO_ENABLED=1`, the binary depends on a specific Glibc version, one that differs between the two containers.
> 
> I think the solution is to build in the same container which cloud-init will run, `rockylinux:9`. I have gotten a successful build working on my branch here: [`synackd/container-build-fixes`](https://github.com/OpenCHAMI/cloud-init/tree/synackd/container-build-fixes). It adds a separate `.goreleaser-buildincontainer.yaml` file that sets appropriate linker variables for ARM, as well as a separate `build-in-container-r9.sh` build script. These were added as separate files so that the differences can be seen with the existing ones.
> 
> With docker:
> 
> ```
> ./build-in-container-r9.sh
> ```
> 
> With Podman (what I've tested with, Podman socket needs to be enable):
> 
> ```
> CONTAINER_CMD=podman CONTAINER_SOCK=/run/podman/podman.sock ./build-in-container-r9.sh
> ```

Upon further discussion, it might be better to change the container that cloud-init runs in from `rockylinux:9` to `ubuntu:24.04` since GitHub actions uses an Ubuntu runner and it's easier to cross compile in Ubuntu than it is in Rocky Linux.

alexlovelltroy commented:
I tried several different things, but switching to an Ubuntu base container makes the CGO stuff work better with the fewest complications.

synackd reviewed (changes requested):
I was attempting to test with `ochami` and was getting a bunch of 404s. Looks like the API got modified in an undocumented way which breaks functionality. See my review comment.

alexlovelltroy commented on Dockerfile:
```diff
@@ -32,8 +13,7 @@ ENV SMD_URL="http://smd:27779"
 # Set non-root user
-# UID 65534 is typically "nobody" on Red Hat systems
-USER 65534:65534
+USER 10001
```
I remembered that as the default UID for nobody in ubuntu containers.  I checked and that's wrong.  Reverting it to 65534.

synackd commented:
When testing, this, I could successfully add and retrieve cluster defaults, but not a cloud-config.

Using the following YAML:

```yaml
---
- name: compute
  description: "compute group ssh config templates"
  file:
    encoding: plain
    content: |
      ## template: jinja
      #cloud-config
      merge_how:
      - name: list
        settings: [append]
      - name: dict
        settings: [no_replace, recurse_list]
      users:
        - name: root
          ssh_authorized_keys: {{ ds.meta_data.instance_data.v1.public_keys }}
      disable_root: false
```

I set the group config:

```
$ ochami cloud-init group set -f yaml -d @/etc/openchami/data/cloud-init/ci-group-compute.yaml --log-level debug --log-format basic
DEBUG | lib.go:177 > logging has been initialized
DEBUG | lib.go:309 > using base URI from default cluster redondo
DEBUG | lib.go:405 > authentication enabled for cluster redondo, reading and checking token
DEBUG | lib.go:437 > Determining token from environment variable based on cluster in config file
DEBUG | lib.go:443 > --cluster not specified, using default-cluster: redondo
DEBUG | lib.go:454 > Reading token from environment variable: REDONDO_ACCESS_TOKEN
DEBUG | lib.go:456 > Token found from environment variable: REDONDO_ACCESS_TOKEN=ey...
DEBUG | client.go:389 > payload file: /etc/openchami/data/cloud-init/ci-group-compute.yaml
DEBUG | client.go:390 > payload file format: yaml
DEBUG | client.go:406 > bytes read: "---\n- name: compute\n  description: \"compute group ssh config templates\"\n  file:\n    encoding: plain\n    content: |\n      ## template: jinja\n      #cloud-config\n      merge_how:\n      - name: list\n        settings: [append]\n      - name: dict\n        settings: [no_replace, recurse_list]\n      users:\n        - name: root\n          ssh_authorized_keys: {{ ds.meta_data.instance_data.v1.public_keys }}\n      disable_root: false\n"
DEBUG | client.go:245 > PUT: https://redondo.usrc:8443/cloud-init/admin/groups/compute
DEBUG | client.go:266 > Request headers:
DEBUG | client.go:268 >   User-Agent: [ochami/0.5.1]
DEBUG | client.go:268 >   Authorization: [Bearer ey...]
DEBUG | client.go:274 > Request body:
DEBUG | client.go:275 > {"name":"compute","description":"compute group ssh config templates","file":{"content":"## template: jinja\n#cloud-config\nmerge_how:\n- name: list\n  settings: [append]\n- name: dict\n  settings: [no_replace, recurse_list]\nusers:\n  - name: root\n    ssh_authorized_keys: {{ ds.meta_data.instance_data.v1.public_keys }}\ndisable_root: false\n","filename":"","encoding":"plain"}}
DEBUG | client.go:288 > Response status: 201 Created
DEBUG | client.go:290 > Response headers:                                                                
DEBUG | client.go:292 >   Location: [/groups/compute]
DEBUG | client.go:292 >   Date: [Mon, 25 Aug 2025 19:17:34 GMT]
DEBUG | client.go:292 >   Content-Length: [0]
DEBUG | client.go:309 > No body in response
INFO  | http.go:124 > Response status: HTTP/2.0 201 Created
```

In the logs, I see that the cloud-config is blank:

```
cloud-init-server[1572728]: Storing data for group compute: {"name":"compute","description":"compute group ssh config templates","file":{"content":"","filename":"","encoding":"plain"}}
cloud-init-server[1572728]: 2025/08/25 19:17:34 [cloud-init/ve6TEoeAmE-000008] "PUT http://redondo.usrc:8443/admin/groups/compute HTTP/1.1" from 172.16.0.254 - 201 0B in 25.615447ms
cloud-init-server[1572728]: 7:17PM INF Request bytes_in=380 bytes_out=0 duration=25.514526 method=PUT remote_addr=172.16.0.254 request_id=cloud-init/ve6TEoeAmE-000008 request_uri=/admin/groups/compute status=Created status_code=201 user_agent=ochami/0.5.1
```

The first line originates here:

https://github.com/OpenCHAMI/cloud-init/blob/d4e26727ea9f380c162c8ccb0d26a2cc84f4b515/internal/quackstore/quackstore.go#L157

which smells like a JSON marshalling issue.

The `GroupData` and `CloudConfigFile` structs appear to have proper JSON reflect tags, so not sure where the problem is occurring yet. My guess is that it has something to do with marshalling the JSON to store in DuckDB. Perhaps something with the `UnmarshalJSON` function?

synackd commented:
Yep, looks to be the `MarshalJSON` function. I committed a fix on my [branch](https://github.com/OpenCHAMI/cloud-init/tree/synackd/quackbackend-fixes). See the diff [here](https://github.com/OpenCHAMI/cloud-init/commit/47d7d19ab4a1dc49f7dbe8396a59260b1d8687e3).

synackd commented:
With the fix, getting/setting defaults and cloud-configs seems to work. I also tested rendering and fetching node meta-data, which both worked. However, I still see:

```
cloud-init-server[1572728]: 7:04PM ERR Failed to load DuckDB extensions error="IO Error: Can't find the home directory at '/nonexistent'\nSpecify a home directory using the SET home_directory='/path/to/dir' option."
```
every time cloud-init starts.

synackd reviewed (changes requested):
There are a _lot_ of places where errors are being ignored, many of which I don't quite understand if there's a legitimate reason or not. If there is a legitimate reason, they should be commented (I did see some comments already) since ignoring errors in production code is generally not a good practice since it makes it hard to troubleshoot where problems are arising.

Besides that, there were a few idiomatic suggestions I added.

This was purely a code review and I still need to test it.

synackd commented:
Other than the DuckDB plugins error still occurring, everything else appears to work.

alexlovelltroy commented:
@synackd Can you confirm that you don't get that error anymore?   I'm building the container differently now so we can preload all plugins which is also safer.

synackd commented:
This is what I see now:
```
cloud-init-server[1952550]: 2025/09/18 20:29:38 failed to create modcache index dir: mkdir /nonexistent: permission denied
...
cloud-init-server[1952550]: 8:29PM ERR Failed to load DuckDB extensions error="IO Error: Failed to download extension \"json\" at URL \"http://extensions.duckdb.org/v1.1.3/linux_amd64/json.duckdb_extension.gz\"\nExtension \"json\" is an existing extension.\n (ERROR Could not establish connection)" home=/opt/duckdbhome
```

synackd commented:
I no longer see the error! Are there any more impending changes that need to be made?

alexlovelltroy commented:
Not that I'm aware of.  Once you approve this PR, we can merge it.

synackd commented on cmd/cloud-init-server/main.go:
```diff
@@ -229,25 +221,30 @@ func startServer() error {
-	// Client sub-router
-	router_client := chi.NewRouter()
-	initCiClientRouter(router_client, ciHandler, wgInterfaceManager)
-	router.Mount("/cloud-init", router_client)
```
Mounting things at `/` instead of `/cloud-init` is a welcome change, but it will be a breaking change to users of the release RPM unless
```
http-request replace-path ^/cloud-init(/.*) \1
```
is added to the cloud-init backend in the HAproxy config.

We can certainly make this change when incorporating this PR into the release repo, but I think it's worth pointing out here for posterity.

synackd reviewed (commented):
I've pushed a commit updating the documentation of the removal of `/cloud-init`. Let me know your thoughts on this. I think this is the last change needed before merging.

synackd commented:
Missed a few things, but commit is up-to-date now.
